### PR TITLE
Add TPSG docs

### PR DESCRIPTION
**Pull request by sivanov-nuodb**

### CODE REVIEW

File: docs/TablePartitionsStorageGroups.md [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+Each partition within a table has a uniquely identified name that maps to one storage group (SG) which represents a logical storage unit that can be serviced by multiple SMs in a database.
+
+This document explains how to configure Table Partitions and Storage Groups (TPSG) in NuoDB Kubernetes deployments.
+
+>**Note**
+> For non-Kubernetes deployments, see  [Table Partitions and Storage Groups][2]. This page expands on the product documentation and is specific to this Helm Chart repository.
+
+## Deployment Overview
+
+NuoDB [database](../stable/database) Helm chart can be installed multiple times in a single Kubernetes namespace where there is one _primary_ and one or more _secondary_ Helm release.
```
Comment (adriansuarez) [inline docs/TablePartitionsStorageGroups.md:44]: You can also have zero secondary releases. I would rephrase this as "... and any number of secondary Helm releases".

File: docs/TablePartitionsStorageGroups.md [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+A storage group is deployed as a separate _secondary_ Helm release which also deploys the SMs servicing this storage group.
+This allows the user to remove any storage group without removing resources installed by the _primary_ database Helm release.
+To simplify TPSG operations and observability, NuoDB Helm charts support one storage group per Storage Manager.
```
Comment (adriansuarez) [inline docs/TablePartitionsStorageGroups.md:47]: Should this say "Storage Manager StatefulSet"?

File: docs/TablePartitionsStorageGroups.md [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+In this example, a NuoDB database with two storage groups is provisioned into the Kubernetes namespace.
+The Transaction Engines (TEs) are part of the _primary_ Helm release which is deployed first.
+One backup group is configured that includes all Hot-copy Storage Managers (HCSMs) as there will be one deployed per storage group.
+For more information on configuring backup when TPSG is enabled, see [Backup Database with TPSG](#backup-database-with-tpsg).
+
+```sh
+helm upgrade --install database stable/database \
+  --set database.te.replicas=3 \
+  --set database.sm.hotCopy.replicas=0 \
+  --set database.sm.noHotCopy.enablePod=false \
```
Comment (adriansuarez) [inline docs/TablePartitionsStorageGroups.md:74]: Is the difference between `database.sm.noHotCopy.enablePod=false` and `database.sm.noHotCopy.replicas=0` that the former suppresses the backup cronjob? 

It might be clearer and more complete to specify `database.sm.noHotCopy.replicas=0` even if it is redundant with `enablePod=false`, since it shows all the workload replica counts.

File: docs/TablePartitionsStorageGroups.md (hunk 1 above)
Comment (sivanov-nuodb) [reply docs/TablePartitionsStorageGroups.md:44]: Will do.

File: docs/TablePartitionsStorageGroups.md (hunk 2 above)
Comment (sivanov-nuodb) [reply docs/TablePartitionsStorageGroups.md:47]: Both are correct but adding "StatefulSet" makes in more precise.

File: docs/TablePartitionsStorageGroups.md (hunk 3 above)
Comment (sivanov-nuodb) [reply docs/TablePartitionsStorageGroups.md:74]: The `database.sm.hotCopy.enablePod=false` also suppresses the CronJob rendering which is unfortunate as there is another variable (`database.sm.hotCopy.enableBackups`) that controls this. We need the CronJobs here, hence, `database.sm.hotCopy.replicas=0` is set. The downside is that the HCSM StatefulSet will also be rendered but it's not a problem.

I will change `database.sm.noHotCopy.enablePod=false` -> `database.sm.noHotCopy.replicas=0` in the primary release so that it's consistent.

**Review by bkelly-ndb — APPROVED** [review]

File: docs/TablePartitionsStorageGroups.md [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+> Scaling down all TEs in the database will make the database unavailable to applications so make sure to schedule this operation.
+
+```sh
+kubectl scale deployment -l domain=nuodb,database=demo,component=te --replicas 0
+```
+
+Start the existing storage group which was previously unavailable.
+
+```sh
+kubectl scale statefulset -l domain=nuodb,database=demo,release=sgeast --replicas 0
```
[the remark this answers is not in the document]
Comment (sivanov-nuodb) [reply docs/TablePartitionsStorageGroups.md:259]: Good catch! Will fix.

File: docs/TablePartitionsStorageGroups.md [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+
+Get the archive IDs that are associated with the storage group.
+The `archive_states` property has all archive IDs associated to this storage group.
+
+```sh
+kubectl exec admin-nuodb-cluster0-0 -- \
+    nuocmd get storage-groups --db-name demo
+```
+
+If [Kubernetes Aware Admin][1] is configured, there will be only one archive associated with the storage group as KAA will automatically remove archives for scaled-down SMs.
```
Comment (adriansuarez) [inline docs/TablePartitionsStorageGroups.md:229]: I tend to use the terms "deprovision" or "disable" when talking about archive objects being removed from the domain state _without_ `purge=true`. Should we make it clear that the archive is not actually being removed, and also that no data associated with it has been deleted?

File: docs/TablePartitionsStorageGroups.md [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+...
+StorageGroup(archive_states={0: ADDED}, db_name=demo, id=10, leader_candidates=[], name=SGEAST, process_states={}, state=Unavailable)
+...
+```
+
+Delete, _but not_ purge, the last archive that services this storage group.
+
+```sh
+kubectl exec admin-nuodb-cluster0-0 -- \
+    nuocmd delete archive --archive-id 0
```
Comment (adriansuarez) [inline docs/TablePartitionsStorageGroups.md:241]: Is this actually necessary? I know that a database cannot be left without archive objects, but there is nothing in the nuoadmin code that prevents the final archive for a storage group other than UNPARTITIONED from being removed. Is this enforced somewhere else, e.g. KAA?

File: docs/TablePartitionsStorageGroups.md [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+```
+
+### Start Unavailable Storage Group
+
+Starting a previously _unavailable_ storage group in a database with running TEs is not supported.
```
Comment (adriansuarez) [inline docs/TablePartitionsStorageGroups.md:246]: Should say "storage groups", plural.

File: docs/TablePartitionsStorageGroups.md [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+All TEs must be scaled down before the previously _unavailable_ storage group can be started.
+
+>**Warning**
+> Scaling down all TEs in the database will make the database unavailable to applications so make sure to schedule this operation.
+
+```sh
+kubectl scale deployment -l domain=nuodb,database=demo,component=te --replicas 0
+```
```
Comment (adriansuarez) [inline docs/TablePartitionsStorageGroups.md:254]: Should we add a note about scaling down all TE groups, if the database has multiple TE groups.

File: docs/TablePartitionsStorageGroups.md [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+
+```sh
+kubectl scale statefulset -l domain=nuodb,database=demo,release=sgeast --replicas 0
+```
+
+Start the TEs which will wait for the storage group to become available.
+
+```sh
+kubectl scale deployment -l domain=nuodb,database=demo,component=te --replicas 3
+```
```
Comment (adriansuarez) [inline docs/TablePartitionsStorageGroups.md:266]: It's possible that the TE groups had different replica counts originally, so I would mention that in the note as well.

File: docs/TablePartitionsStorageGroups.md (hunk 6 above)
Comment (sivanov-nuodb) [reply docs/TablePartitionsStorageGroups.md:241]: The last archive for a storage group won't be automatically deleted (there is an enforcement in KAA). It should be deleted manually so that `nuodocker` doesn't wait for that storage group to become _available_ anymore on the database cold start. A storage group without archives will be excluded from the covering set of storage groups for the database during hot-copy as well.

File: docs/TablePartitionsStorageGroups.md [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+
+### Backup Failure
+
+The output below shows a failed _full_ backup:
+
+```text
+Executing full hotcopy as a prerequisite for incremental hotcopy: full hotcopy in backupset 20230919T053410 has failed
+Starting full backup for database demo on processes with labels 'role hotcopy' ...
+'backup database' failed: Failure while performing hot-copy: Not all storage groups are included in hot-copy for dbName=demo, missingSgs=[SGEAST]
+Error running hotcopy 1
```
Comment (adriansuarez) [inline docs/TablePartitionsStorageGroups.md:389]: This "Error running hotcopy 1" line is coming from `kubectl exec`, right? I think you can drop it since it is not clear that you're executing this using `kubectl exec`.

Comment (sivanov-nuodb) [reply docs/TablePartitionsStorageGroups.md:389]: It is from `nuobackup` output. Actually, I can leave just the line indicating the error and strip the rest to avoid confusion.

File: docs/TablePartitionsStorageGroups.md (hunk 9 above)
Comment (sivanov-nuodb) [reply docs/TablePartitionsStorageGroups.md:266]: Will do.

File: docs/TablePartitionsStorageGroups.md (hunk 5 above)
Comment (sivanov-nuodb) [reply docs/TablePartitionsStorageGroups.md:229]: Good point.
